File: differenza_chiusura_relazioni.hs
{- Programma Haskell per calcolare differenze tra relazioni e chiusure di relazioni -}

import Data.List -- necessario per usare nub, che elimina elementi duplicati da una lista

type Rel = [(Int, Int)]

main :: IO ()
main = do putStrLn "Inserisci l'insieme di numeri naturali racchiuso tra parentesi quadre:"
          insStr <- getLine
          let insieme = nub( read insStr :: [Int])
          putStr "Inserire la relazione binaria sull'insieme racchiusa tra parentesi quadre:"
          relazioneStr <- getLine
          let relazione = nub (read relazioneStr :: Rel)
          putStrLn "\nChiusura riflessiva di R1:"
          putStrLn $ show (riflessiva insieme r1)
          putStrLn "\nChiusura simmetrica di R2:"
          putStrLn $ show (simmetrica r2)

{- La funzione riflessiva calcola la chiusura riflessiva di una relazione:
   - il primo argomento è il codominio della relazione, ovvero l'insieme di numeri interi inserito
   - il secondo argomento è la relazione stessa -}

riflessiva :: [Int] -> Rel -> Rel
riflessiva [] rel = rel   
riflessiva (x:xs) rel =
    let coppia = (x, x)                 
        rel'   = if elem coppia rel   
                    then rel
                    else coppia : rel
    in riflessiva xs rel'       

{- La funzione simmetrica calcola la chiusura simmetrica di una relazione, 
   - il suo unico argomento è la relazione stessa -}

simmetrica :: Rel -> Rel
simmetrica [] = []                                
simmetrica ((a,b):xs) =
    let resto = simmetrica xs                     
        coppie = if elem (b,a) (a,b):xs         
                    then [(a,b)]                  
                    else [(a,b),(b,a)]            
    in nub (coppie ++ resto)                      

File: differenza_chiusura_relazioni.pl
% differenza tra due relazioni
diff([], _, []).
diff([X|Xs], Y, R) :-
    member(X, Y), !,
    diff(Xs, Y, R).
diff([X|Xs], Y, [X|R]) :-
    diff(Xs, Y, R).

% chiusura riflessiva: aggiunge (X,X) per ogni X dell’insieme
riflessiva([], R, R).
riflessiva([X|Xs], R, RR) :-
    (member((X,X), R) -> R1 = R ; R1 = [(X,X)|R]),
    riflessiva(Xs, R1, RR).

% chiusura simmetrica
simmetrica([], R, R).
simmetrica([(A,B)|Xs], R, RR) :-
    (member((B,A), R) -> R1 = R ; R1 = [(B,A)|R]),
    simmetrica(Xs, R1, RR).

% main di esempio
run :-
    writeln('Inserisci l\'insieme (lista es. [1,2,3]).'),
    read(Insieme),
    writeln('Inserisci la prima relazione (es. [(1,2),(2,3)]).'),
    read(R1),
    writeln('Inserisci la seconda relazione (es. [(2,3),(3,1)]).'),
    read(R2),

    diff(R1, R2, DR1),
    writeln('Differenza R1 \\ R2:'), writeln(DR1),

    diff(R2, R1, DR2),
    writeln('Differenza R2 \\ R1:'), writeln(DR2),

    riflessiva(Insieme, R1, Rif),
    writeln('Chiusura riflessiva di R1:'), writeln(Rif),

    simmetrica(R2, R2, Sim),
    writeln('Chiusura simmetrica di R2:'), writeln(Sim).